### ISSUE TITLE
Some widgets not working in Ubuntu: TaskList, WindowTabs, CheckUpdates

### ISSUE THREAD

**Issue by soulmare** [qtile/qtile#1332]
Some widgets are not working in Ubuntu at all. When enabling such a widget in config without any parameters, Qtile seems to restart, but panel disappears and all key bindings stop working. And nothing special is firing to the qtile.log, even if messages level is set to DEBUG.
Widgets having this problem:
- TaskList
- WindowTabs
So, I have a question: is there a (simple) way to realize what breaks Qtile in these cases? _I just want to see an error text with filename and line number_. Log file does not give any info. Can I turn on viewing errors somehow?
Also, CheckUpdates always shows "N/A", but I know some updates must be available right now.
Ubuntu 18.04, Qtile 0.13

**Comment by ramnes** [comment]
Does `python config.py` give any output? :slightly_smiling_face:

**Comment by soulmare** [comment]
Only notices that some optional libs are missing (wifi etc)

**Comment by soulmare** [comment]
I'm running qtile in python3 btw

**Comment by ramnes** [comment]
Can you share a minimal example configuration that makes this crash happens? I can't reproduce this (and always have logs in `~/.local/share/qtile/qtile.log` when something is going wrong).

**Comment by tych0** [comment]
Also, check ~/.xsession-errors if things crash really early on.

**Comment by soulmare** [comment]
Tested my issues in virtual machine with clean Ubuntu 18.04. Installed Qtile 0.10 from Ubuntu repos, used a default config, and after adding TaskList or WindowTabs to the widgets list, it crashes Qtile. In qtile.log there are messages that Qtile is crashed, and **only one STRETCH widget is allowed** in one time. So, I've fixed my config disabling conflicting WindowName widget, and TaskList now works. However, on my machine I have Qtile 0.13 installed via pip3, and **with same config file it just silently crashes with no message to the log file**.
I think, it would be more user-convenient in such case **to crash only a conflicting widget, not the whole Qtile**.

Also, in that virtual machine I had some minor issues:
- Qtile config file does not exist - searching for config.py gives no relevant results. Don't know, is this normal or not.
- Terminal keybind is not working, because there's no xterm in Ubuntu by default (gnome-terminal is used instead). Maybe, it would be nice to get terminal name from environment variable or guess it somehow, not hardcoding xterm by default. You just install qtile on Ubuntu, and what you want to do after first started it, is launching a couple of terminals, but it just doesn't work out of the box :(

**Comment by tych0** [comment]
Without logs, ti's tough to say what happened. Can you check
~/.xsession-errors too?

**Comment by soulmare** [comment]
For fresh installed Ubuntu 18.04:

~/.xsession-errors does not exist.

~/.local/share/qtile/qtile.log:

`324 ERROR libqtile qtile.py:make_qtile():L119 Qtile crashed during startup
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/libqtile/scripts/qtile.py", line 116, in make_qtile
    state=options.state,
  File "/usr/lib/python3/dist-packages/libqtile/manager.py", line 175, in __init__
    self._process_screens()
  File "/usr/lib/python3/dist-packages/libqtile/manager.py", line 383, in _process_screens
    self.groups[i],
  File "/usr/lib/python3/dist-packages/libqtile/config.py", line 268, in _configure
    i._configure(qtile, self)
  File "/usr/lib/python3/dist-packages/libqtile/bar.py", line 181, in _configure
    raise confreader.ConfigError("Only one STRETCH widget allowed!")
libqtile.confreader.ConfigError: Only one STRETCH widget allowed!
`

**Comment by tych0** [comment]
Sure, that is a legitimate error. I'm interested in the second one, where you have 0.13.0; I suspect that's related to some kind of name that has changed.

**Comment by soulmare** [comment]
For the case with 0.13.0, installed via "pip3 install", when added to default config TaskList widget, Qtile crashes and I see in ~/.local/share/qtile/qtile.log:
> 2019-06-04 08:58:59,353 WARNING libqtile manager.py:finalize():L355 Restarting Qtile with os.execv(...)
2019-06-04 08:58:59,642 WARNING libqtile log_utils.py:init_log():L120 Starting logging for Qtile
2019-06-04 08:58:59,763 WARNING libqtile utils.py:safe_import():L268 Unmet dependencies for optional Widget: '.widget.mpdwidget.Mpd', No module named 'mpd'
2019-06-04 08:58:59,765 WARNING libqtile utils.py:safe_import():L268 Unmet dependencies for optional Widget: '.widget.mpd2widget.Mpd2', No module named 'mpd'
2019-06-04 08:58:59,770 WARNING libqtile utils.py:safe_import():L268 Unmet dependencies for optional Widget: '.widget.wlan.Wlan', No module named 'iwlib'
2019-06-04 08:58:59,857 WARNING libqtile xcursors.py:_setup_xcursor_binding():L117 xcb-cursor not found, fallback to font pointer

That's it. This log is just the same as is generated on normal start with not modified config file.
~/.xsession-errors does not exist.
python3 config.py also gives me the same result with both configs, original and modified:
> Unmet dependencies for optional Widget: '.widget.mpdwidget.Mpd', No module named 'mpd'
Unmet dependencies for optional Widget: '.widget.mpd2widget.Mpd2', No module named 'mpd'
Unmet dependencies for optional Widget: '.widget.wlan.Wlan', No module named 'iwlib'

I believe this issue is easy reproduceable: just did it on VM with fresh installed Ubuntu 18.04 (used VM image downloaded from osboxes.org). Updated OS, installed all dependencies, then pip3 install qtile, then added one line to default config, that's all.

**Comment by daroczig** [comment]
I can confirm this issue :confused: 

This started to happen on my end when update `bar.Bar` to be in the `top` of the `Screen` -- at the default `bottom` it worked OK. But in the `top`, adding either `widget.WindowName()` or eg `widget.WindowTabs(separator="    |    ", **style)` fails Qtile (the top bar disappears, no keybindings working after restarting Qtile) -- just like @soulmare reported.

**Comment by pawamoy** [comment]
+1

Top bar disappears when I use Spacer or TaskList. I suspected that multiple "spacers" could be the cause, but even removing SysTray and using only TaskList causes QTile to crash.

Spacer in the bottom bar works fine however, though I did not try with combinations of SysTray and/or TaskList.

**Comment by richardtatum** [comment]
I am not sure if I am reviving a old issue here but I am still having this problem as of 0.15.1-1. I am also on Arch Linux.

If I enable the TaskList I loose the top bar completely and none of keybindings work. 

Snippet from my `config.py`
```
                 widget.Spacer(length=5),
                 widget.WindowName(fontsize=12),
                 widget.Spacer(length=5),
                 widget.TaskList(),
                 widget.Spacer(length=5),
                 widget.Notify(default_timeout=10),
                 widget.Systray(icon_size=22, padding=10),
```

Running the file first with Python shows no errors.
I cannot restart qtile with `qtile-cmd -o cmd -f restart` as I get the same errors as #1355 and qtile does not restart.

The only thing of note in the logs is the following:

```
2020-04-24 16:46:06,183 ERROR libqtile manager.py:<lambda>():L200 Got an exception in poll loop
Traceback (most recent call last):
  File "/usr/lib/python3.8/asyncio/selector_events.py", line 872, in _read_ready__on_eof
    keep_open = self._protocol.eof_received()
  File "/usr/lib/python3.8/site-packages/libqtile/ipc.py", line 282, in eof_received
    self.transport.write(result)
  File "/usr/lib/python3.8/asyncio/selector_events.py", line 893, in write
    raise RuntimeError('Cannot call write() after write_eof()')
RuntimeError: Cannot call write() after write_eof()
2020-04-24 16:46:06,188 WARNING libqtile manager.py:loop():L245 Restarting Qtile with os.execv(...)
```
But again I believe this to be linked with the restart failing.

If I can provide anything else (or if you would prefer me to open a new issue) please let me know.

**Comment by CSpencerND** [comment]
Any progress on this issue? I can't get TaskList to work either. My bar disappears and I get no error messages.

**Comment by elParaguayo** [comment]
What version of qtile are you running? Is this still version 0.15? If so, you need to upgrade to the latest version.

**Comment by CSpencerND** [comment]
0.18.1

**Comment by elParaguayo** [comment]
I wonder if this is to do with the icons. What happens if you set `icon_size=0`? does it still crash?

**Comment by CSpencerND** [comment]
Wow, that's interesting. `icon_size=0` makes it work.
```
widgets_list = [
    widget.CurrentLayoutIcon(
        custom_icon_paths=[path + 'layouts'],
        scale=0.7
    ),
    widget.TaskList(icon_size=0),
    widget.Clock(format='%Y-%m-%d %a %I:%M %p'),
    widget.Systray(),
]

screens = [
    Screen(top=bar.Bar(widgets_list, 24, opacity=0.95))
]
```

**Comment by elParaguayo** [comment]
It was just a hunch. The rest of the icons is pretty straightforward so I didn't think that would be the problem.

So, given we know it's the icons, let's see if we can find an error. Let's try this, open a number of windows (ideally those that you had open when the widget crashed), then open a python shell and run the following:
``` bash
>>> from libqtile.command.client import InteractiveCommandClient
>>> i=InteractiveCommandClient()
>>> i.eval("[win.icons for win in self.current_group.windows]")
```

If there was an error there, I would have thought this should have ended up in the logs but this will still help track down the problem.

**Comment by CSpencerND** [comment]
That gave me a gigantic list of integers, but now I'm suddenly getting errors in qtile.log
```
2021-08-12 01:15:57,824 ERROR libqtile loop.py:_handle_exception():L59 ffi_prep_closure(): bad user_data (it seems that the version of the libffi library seen at runtime is different from the 'ffi.h' file seen at compile-time)
Traceback (most recent call last):
  File "/usr/lib/python3.9/asyncio/events.py", line 80, in _run
    self._context.run(self._callback, *self._args)
  File "/home/cs/.local/lib/python3.9/site-packages/libqtile/core/manager.py", line 792, in f
    func(*args)
  File "/home/cs/.local/lib/python3.9/site-packages/libqtile/bar.py", line 440, in _actual_draw
    i.draw()
  File "/home/cs/.local/lib/python3.9/site-packages/libqtile/widget/tasklist.py", line 469, in draw
    for w, icon, task, bw in self.calc_box_widths():
  File "/home/cs/.local/lib/python3.9/site-packages/libqtile/widget/tasklist.py", line 266, in calc_box_widths
    icons = [self.get_window_icon(w) for w in windows]
  File "/home/cs/.local/lib/python3.9/site-packages/libqtile/widget/tasklist.py", line 266, in <listcomp>
    icons = [self.get_window_icon(w) for w in windows]
  File "/home/cs/.local/lib/python3.9/site-packages/libqtile/widget/tasklist.py", line 431, in get_window_icon
    img = cairocffi.ImageSurface.create_for_data(
  File "/usr/lib/python3/dist-packages/cairocffi/surfaces.py", line 743, in create_for_data
    return cls(format, width, height, data, stride)
  File "/usr/lib/python3/dist-packages/cairocffi/surfaces.py", line 735, in __init__
    Surface.__init__(self, pointer, target_keep_alive=data)
  File "/usr/lib/python3/dist-packages/cairocffi/surfaces.py", line 164, in __init__
    keep_alive = KeepAlive(target_keep_alive)
  File "/usr/lib/python3/dist-packages/cairocffi/surfaces.py", line 117, in __init__
    callback = ffi.callback(
SystemError: ffi_prep_closure(): bad user_data (it seems that the version of the libffi library seen at runtime is different from the 'ffi.h' file seen at compile-time)

```

**Comment by elParaguayo** [comment]
OK. Now we're getting somewhere.

The big list of integers is correct but that error is almost certainly the source of your problems.

It sounds like some of your python packages may be out of date, probably `cffi`.

How did you install qtile?

**Comment by CSpencerND** [comment]
I installed Qtile from source. After reading those errors I then installed cffi, and libffi from source. Then I ran locate to find ffi.h and found it in three different places. I replaced  the two that didn't match libffi version 3.4.2 ( the newest)  with the 3.4.2 version. Still getting the same error after all that. You think I should remove all that stuff and install them fresh?

**Comment by elParaguayo** [comment]
OK. Presumably you followed the guide [here](https://docs.qtile.org/en/latest/manual/install/index.html#installing-from-source) (noting the order about installing `xcffib` before `cairocffi`)?

If so, then I may need to reach out to some of the team who are more familiar with ffi then me!

**Comment by CSpencerND** [comment]
At this point, I usually chalk it up to Ubuntu being Ubuntu and move on. Ha. But yes. I remember clearly following those instructions carefully, but human error is always a thing. I could try reinstalling them just to be sure. You think it might break something if I tried that?

**Comment by elParaguayo** [comment]
I wouldn't have thought so but @tych0 is much more familiar with the ffi stuff so may have a better idea as to why you're getting that error.

**Comment by tych0** [comment]
I would install exactly one libffi (any of the Ubuntu packaged version or your locally built version would be fine, but there should be only one), clear your entire pip cache (which can be hard to do, so be sure you got it all), then rebuild everything and try again.

It is likely possible to fix it using the right LD_LIBRARY_PATH, etc. but I probably wouldn't bother with that unless I really wanted a custom libffi.

**Comment by CSpencerND** [comment]
Gotcha. Idk how the heck I have three. I'll get this sorted out. Thanks for all the help, guys. I'll let ya know what happens. :)

**Comment by CSpencerND** [comment]
So, here's what happened. I removed all the libffi's. Reinstalled only the newest version. Logged out of Qtile and GDM wouldn't load. Got the "Something went wrong message." It seems GDM required some other version of libffi. Couldn't startx from a tty either. Did some more research. I installed the version of libffi that GDM needed. Tried again. GDM still doesn't work and even worst, it locks up and I can't even get a tty. So ya know what? Fuggit. I installed Arch and everything works fine XD.
